src/migrate.c

```diff
@@ -0,0 +1,171 @@
+#include <string.h>
+
+#include "redisraft.h"
+
+static int validSlotMagic(RedisRaftCtx *rr, unsigned int slot, int magic)
+{
+    /* validSlot will have already validated that slot is importing */
+    ShardGroup *sg = rr->sharding_info->importing_slots_map[slot];
+
+    for (unsigned int i = 0; i < sg->slot_ranges_num; i++) {
+        ShardGroupSlotRange *sr = &sg->slot_ranges[i];
+        if (sr->start_slot <= slot && sr->end_slot >= slot && sr->magic == magic) {
+            return 1;
+        }
+    }
+
+    return 0;
+}
+
+static int validSlotTerm(RedisRaftCtx *rr, int slot, raft_term_t term)
+{
+    ShardingInfo *si = rr->sharding_info;
+
+    if (si->max_importing_term[slot] <= term) {
+        si->max_importing_term[slot] = term;
+        return 1;
+    }
+
+    return 0;
+}
+
+static int validSlot(RedisRaftCtx *rr, int slot)
+{
+    ShardGroup *sg = rr->sharding_info->importing_slots_map[slot];
+
+    if (sg && sg->local) {
+        return 1;
+    }
+
+    return 0;
+}
+
+void importKeys(RedisRaftCtx *rr, raft_entry_t *entry)
+{
+    RedisModule_Assert(entry->type == RAFT_LOGTYPE_IMPORT_KEYS);
+
+    RaftReq *req = entry->user_data;
+
+    ImportKeys import_keys = {0};
+    RedisModule_Assert(RaftRedisDeserializeImport(&import_keys, entry->data, entry->data_len) == RR_OK);
+    RedisModule_Assert(import_keys.num_keys > 0);
+
+    // FIXME: validate no cross slot migration at append time
+    int slot = KeyHashSlotRedisString(import_keys.key_names[0]);
+
+    if (!validSlot(rr, slot)) {
+        if (req) {
+            RedisModule_ReplyWithError(req->ctx, "ERR not an importing slot");
+        }
+        goto exit;
+    }
+
+    if (!validSlotMagic(rr, slot, import_keys.magic)) {
+        if (req) {
+            RedisModule_ReplyWithError(req->ctx, "ERR invalid magic");
+        }
+        goto exit;
+    }
+
+    if (!validSlotTerm(rr, slot, import_keys.term)) {
+        // this is acceptable (suspended old leader
+        LOG_WARNING("ignoring import keys as old term");
+        if (req) {
+            RedisModule_ReplyWithError(req->ctx, "ERR invalid term");
+        }
+        goto exit;
+    }
+
+    /* 2  'static' strings need for restore command */
+    RedisModuleString *zero = RedisModule_CreateString(rr->ctx, "0", 1); /* ttl of zero */
+    RedisModuleString *replace = RedisModule_CreateString(rr->ctx, "REPLACE", 1); /* overwrite on import */
+
+    for (size_t i = 0; i < import_keys.num_keys; i++) {
+        RedisModuleString * temp[4];
+        temp[0] = import_keys.key_names[i];
+        temp[1] = zero;
+        temp[2] = import_keys.key_serialized[i];
+        temp[3] = replace;
+
+        enterRedisModuleCall();
+        RedisModuleCallReply *reply;
+        RedisModule_Assert((reply = RedisModule_Call(rr->ctx, "restore", "v", temp, 3)) != NULL);
```

Comment:
I think it's good practice to separate actual statements from assert, since theoretically asserts can be filtered out (even if not currently supported by `RM_Assert`). Saw this in several places.

Reply:
this goes to a different Q, what error checking / validation should we do around RM_Calls().

i.e. RM_Call should never return a null reply if we give it a valid command, right?  we would always have the command itself error out on us.  And how much checking of the output command do we have to do.  in soe cases I'm validating all parts of the output when we get what is essentially a data structure back (array of arrays...)

Reply:
made an issue for this